source/frontend/src/telemetry/metrics.ts

```diff
@@ -0,0 +1,82 @@
+import { metrics } from '@opentelemetry/api';
+import {
+  ATTR_HTTP_RESPONSE_STATUS_CODE,
+  METRIC_HTTP_CLIENT_REQUEST_DURATION,
+  METRIC_HTTP_CLIENT_RESPONSE_BODY_SIZE,
+} from '@opentelemetry/semantic-conventions/incubating';
+
+import { TelemetryConfig } from './config';
+import { isBlocked, isBrowserEnvironment, NETWORK_METER } from './utils';
+
+export const registerNetworkMetrics = (config: TelemetryConfig) => {
+  if (!isBrowserEnvironment() || !window.performance) {
+    return;
+  }
+
+  if (config.debug) {
+    console.info('[INFO] Registering network request metrics');
+  }
+
+  const meter = metrics.getMeter(NETWORK_METER);
+
+  // measure the time it takes to make an XHR request
+  const timeSpentMetric = meter.createHistogram(METRIC_HTTP_CLIENT_REQUEST_DURATION, {
+    description: 'The duration of an outgoing HTTP request.',
+    unit: 's',
+    advice: {
+      explicitBucketBoundaries: [
+        0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10,
+      ],
+    },
+  });
+
+  // collect metrics regarding the payload size of network requests
+  const sizeMetric = meter.createGauge(METRIC_HTTP_CLIENT_RESPONSE_BODY_SIZE, {
+    description: 'Size of network response',
+    unit: 'byte',
+  });
+
+  // Use PerformanceObserver Browser API to track these things
+  // https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver
+  const observer = new PerformanceObserver(list => {
+    const entries = list.getEntriesByType('resource');
+    entries.forEach((entry: PerformanceResourceTiming) => {
+      if (
+        'initiatorType' in entry &&
+        (entry.initiatorType === 'xmlhttprequest' || entry.initiatorType === 'fetch')
+      ) {
+        const uri = new URL(entry.name);
```

Comment:
presumably entry.name is always valued?

Reply:
I can add a null check to be more defensive there but according to https://developer.mozilla.org/en-US/docs/Web/API/PerformanceEntry/name this is meant to be defined